### PR TITLE
feat: add lesson player advanced options

### PR DESCRIPTION
**Pull request by jiejasonliu**
🎉🎉🎉  WE CAN FINALLY SHOW THE OUTCOME OF THE CARD THE MOMENT IT FLIPS 🎉🎉🎉  
architecture (bc code is hard): https://drive.google.com/file/d/1_LHSAmWxgwO3G6dp9qBKmZyPwdtLoR3a/view?usp=sharing

### studying / lesson player control
- move control into its own style unit
- add options to control speech recognition (+ pause length, etc.) and sound effects
- all settings are serialized to local storage, it will persist on reload
- all settings can be hot-swapped/changed during a lesson and will apply its effects almost immediately
- correctness is based off a substring of the spoken text not a perfect match

### styling / responsiveness
- player control as its own styled unit
- player control changes layout on smaller dimensions to support larger volume slider
- popups have the sidebar in mind, now centers itself in the content

### code quality
- refactor some animations to `FadeReveal` which deals with revealing contents in a "unraveling reveal" fashion 
- speech recognition workarounds are now all contained in its respective hook `use-speech-recognition`
- card player hook `use-play-card-audio` no longer has any knowledge of a lesson
  - the only thing it knows is playing an audio file with control over pausing it
- lesson related logic is all hoisted to `use-play-lesson`
  - we are keeping pause, play, and skipping by introducing lifecycles to this hook
  - we do different things based off the lifecycle we are at (e.g. pausing is different when capturing speech vs. playing audio) 
- decided to fix all compile warnings (as any, unused things)

---

#### (responsiveness in popups and the study page)
https://user-images.githubusercontent.com/29434693/199850869-39dd3311-7258-4306-abca-8916219e72ef.mp4

#### 🔊 VOLUME UP 🔊 -- (running through a quick lesson) -- 🔊 VOLUME UP 🔊 
https://user-images.githubusercontent.com/29434693/199851086-5f81b947-aeae-4078-a6b9-7d9e2748eec4.mp4

#### (showing that everything serializes on page refresh/reload)
https://user-images.githubusercontent.com/29434693/199851170-4be20cc8-f5bc-490a-88eb-480fbac345ee.mp4

### CODE REVIEW

File: src/components/audio-control-bar/audio-control-bar.test.tsx [hunk 1]
```diff
@@ -1,12 +1,18 @@
 import React from 'react';
-import { render, screen } from '@testing-library/react';
+import { screen } from '@testing-library/react';
 import userEvent from '@testing-library/user-event';
+import { renderWithTestRoots } from '@/app.test';
 import { noop } from '@/helpers/func';
 import { AudioControlBar } from './audio-control-bar';
 
 describe('StudyFlashcard', () => {
```
Comment (derricklee43) [inline src/components/audio-control-bar/audio-control-bar.test.tsx:8]: I think I forgot to update the test suite name. Could you update that for me?

**Review by derricklee43 — APPROVED** [review]
This looks so good! I'm so impressed you were able to get it all working with the skip and previous buttons!

File: src/components/audio-control-bar/audio-control-bar.test.tsx (hunk 1 above)
Comment (jiejasonliu) [reply src/components/audio-control-bar/audio-control-bar.test.tsx]: done! also added a test for the volume slider

**Comment by jiejasonliu** [comment]
> This looks so good! I'm so impressed you were able to get it all working with the skip and previous buttons!

ty, i really wanted to keep them! and i can finally get some beauty sleep now 😴

**Pull request merged**